### PR TITLE
Modified comparisons to pass the Turkey test.

### LINKED ISSUE #2228: Several tests for nunit fails when run in a development environment with certain regional settings
My development PC has its regional settings set to Swedish. When running `.\build -t test` in the nunit repository I get 7 test failures in each project. Worse yet, when switching to Turkish regional settings I get 48 test failures in the 4.5/4.0 projects, 46 test failures in the 3.5/2.0 projects and another 16 in the netstandard projects.

For now, this is only a problem for contributors that do not have regional settings that are English (or any invariant-like setting), and even then it's an easy fix - just change your regional settings to "English (U.S.)"  and re-run the tests.

I assume that the test suite should not have a dependency on the regional settings of the computer it's running on.

Is it safe to say that all output from nunit should be English/invariant? Some of the test failures are likely the result of the code not being explicit about culture in calls to `.ToString()` and its culture-dependent brethren. I'd guess that that culture in most (all?) cases should be `CultureInfo.InvariantCulture`.

I'm happy to help, if you so wish.

### PR DESCRIPTION
**Pull request by Lette**
This PR fixes #2228.

Please review it carefully. All tests are passing, as they should. I'm just a little bit worried that we might be missing a test that would expose something that would break some code, somewhere in another culture...

### CODE REVIEW

File: src/NUnitFramework/tests/Internal/PlatformDetectionTests.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ public class PlatformDetectionTests
                 bool shouldPass = false;
 
                 foreach( string platform in expected )
-                    if ( shouldPass = platform.ToLower() == testPlatform.ToLower() )
+                    if (shouldPass = platform.ToUpperInvariant() == testPlatform.ToUpperInvariant())
```
Comment (jnm2) [inline src/NUnitFramework/tests/Internal/PlatformDetectionTests.cs]: I think `testPlatform.Equals(platform, StringComparison.InvariantCultureIgnoreCase)` is superior because it states the actual goal and also doesn't allocate and transform temporary strings. Is there a reason we can't do this?

File: src/NUnitFramework/framework/Internal/StringUtil.cs [hunk 2]
```diff
@@ -41,8 +41,10 @@ public class StringUtil
         /// strB is sorted first</returns>
         public static int Compare(string strA, string strB, bool ignoreCase)
         {
-            var comparison = ignoreCase ? StringComparison.CurrentCultureIgnoreCase : StringComparison.CurrentCulture;
-            return string.Compare(strA, strB, comparison);
+            string s1 = ignoreCase ? strA.ToUpperInvariant() : strA;
+            string s2 = ignoreCase ? strB.ToUpperInvariant() : strB;
+
+            return String.CompareOrdinal(s1, s2);
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/StringUtil.cs]: What breaks if we implement this as `string.Equals(strA, strB, ignoreCase ? StringComparison.InvariantCultureIgnoreCase : StringComparison.InvariantCulture)`?

File: src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs [hunk 3]
```diff
@@ -330,16 +330,16 @@ private bool DictionaryEntriesEqual(DictionaryEntry x, DictionaryEntry y, ref To
 
         private bool StringsEqual(string x, string y)
         {
-            string s1 = caseInsensitive ? x.ToLower() : x;
-            string s2 = caseInsensitive ? y.ToLower() : y;
+            string s1 = caseInsensitive ? x.ToUpperInvariant() : x;
+            string s2 = caseInsensitive ? y.ToUpperInvariant() : y;
 
             return s1.Equals(s2);
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs]: Same [What breaks if we implement this as `string.Equals(s1, s2, caseInsensitive ? StringComparison.InvariantCultureIgnoreCase : StringComparison.InvariantCulture)`?]

**Comment by jnm2** [comment]
@nunit/framework-team Please, everyone take a look at this.  I don't know the rationale behind when comparisons should be invariant culture vs current culture vs ordinal.

| | Current (/ ignore case) |
|-|-------------------------|
| `NUnitEqualityComparer.StringsEqual`, `NUnitEqualityComparer.CharsEqual` | `Ordinal`/`CurrentCultureIgnoreCase` |
| `StringUtil.Compare` | `CurrentCulture`/`CurrentCultureIgnoreCase` |
| `PlatformHelper.IsPlatformSupported` | `CurrentCulture` |

| | Proposed (/ ignore case) |
|-|--------------------------|
| `NUnitEqualityComparer.StringsEqual`, `NUnitEqualityComparer.CharsEqual` | `Ordinal`/`InvariantCultureIgnoreCase` |
| `StringUtil.Compare` | `Ordinal`/`InvariantCultureIgnoreCase` |
| `PlatformHelper.IsPlatformSupported` | `InvariantCultureIgnoreCase` |

File: src/NUnitFramework/tests/Internal/PlatformDetectionTests.cs (hunk 1 above)
Comment (Lette) [inline src/NUnitFramework/tests/Internal/PlatformDetectionTests.cs]: Your example works well, and that overload of `string.Equals` is in `netstandard13` so we're good. Also, all tests still pass - even while on Turkish regional settings.

My attempt only changed as little code as possible, but yours id probably better.

File: src/NUnitFramework/framework/Internal/StringUtil.cs (hunk 2 above)
Comment (Lette) [inline src/NUnitFramework/framework/Internal/StringUtil.cs]: Well, `String.CompareOrdinal` returns an `int` while `String.Equals` returns a `bool`, so it'll break compilation... :-)

File: src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs (hunk 3 above)
Comment (Lette) [inline src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs]: Unfortunately, `netstandard` doesn't support those `StringComparison` values.

**Comment by Lette** [comment]
> Also, I'd feel better if this PR came with new tests that fail before your changes and succeed after your changes.

That's the thing about this issue. There actually were failing tests (at most 48, in the 4.0 and 3.5 projects), but only when run on a computer with non-English/non-invariant settings.

So, indeed, I had failing tests before I made the changes and no failing tests afterwards.

**Comment by jnm2** [comment]
@Lette this is difficult though because we need something to fail in CI and even better, on ever contributor's computer without them changing culture settings. I guess we'd need to set up an integration test for that to work.

I'm starting to doubt that invariant-culture comparisons are really what we want. Perhaps the real fix for this needs to be `[SetCulture("en")]` on all the tests that were failing, keeping current-culture comparisons everywhere.

**Comment by ChrisMaddock** [comment]
I haven't dug into this too deeply, but I think there's probably a mix here? At a guess, PlatformHelper should probablty be using Invariant culture, but NUnitEqualityComparer should be using the test's current culture?

From what I gather, comparisons within a test should run based on the Test's current culture (which may have been set to something different from the machine's default, via the SetCulture attribute.). Actual test framework code (e.g. the PlatformAttribute, which the changes to PlatformHelper here are affecting) uses Invariant Culture (Based on https://github.com/nunit/nunit/issues/2008 - although I'm not sure I personally agree with the case specified there!)

I agree this should have tests, if nothing else, there should be something to stop regression - as I doubt many of our developers will double check Turkish Culture before pushing! 😄 It would be good to have duplicate Turkish culture tests using `[SetCulture("turkey")]` where the culture genuinely broke NUnit code. I imagine in other instances - the bug will actually be in our tests, rather than the NUnit code.

I think it might be worth waiting for an authoritative response from @rprouse on this one - I imagine we may end up with a lot of back and forth otherwise, as it's not totally clear what we would want in each culture!

**Comment by Lette** [comment]
A couple of the tests that were failing, failed because they compared `"Unix"` to `"UNIX"` and `"Universe"` to `"UNIVERSE"` by ignoring case through `.ToLower()` (culture-dependent, that is) and expected them to be equal. That is *not* the case in Turkish, since `"i"` is *not* the lower case of `"I"`.

I found [this page on MSDN](https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call) that describes when to use what `StringComparison`member. Alas, all of them are not available in `netstandard`.

**Comment by jnm2** [comment]
@Lette But if you wrote a test with `[SetCulture("tr-TR")]`, I'm thinking you would want `"I"` and `"i"` to be _inequal_ with a case-insensitive assertion. If you force invariant-culture comparisons, that can't happen. We should have a test on this.

I'm thinking the test with the work `UNIVERSE` needs `[SetCulture("en")]`, that's all.

**Comment by Lette** [comment]
One of the places in the production code where I do believe we have a faulty behavior is with the `PlatformAttribute` . If I wrote a test on my Turkish computer that had the `[Platform("Unix")]` attribute, it wouldn't get picked up, since the `PlatformHelper.IsPlatformSupported` method does a `.ToUpper()` on the argument and then compares it to "UNIX", which will not be equal.

But I agree that there might be places where we only should be explicit with the culture of the test itself.

**Comment by jnm2** [comment]
I agree about platform names.

Can you add this? I'm pretty sure we can all agree on this, but it requires current-culture comparisons to be used instead of invariant:

```c#
[Test]
[SetCulture("tr-TR")]
public void TurkishI()
{
    Assert.That("I", Is.Not.EqualTo("i").IgnoreCase);
}

[Test]
[SetCulture("en")]
public void EnglishI()
{
    Assert.That("I", Is.EqualTo("i").IgnoreCase);
}
```

**Comment by Lette** [comment]
I'd love to. But where?

**Comment by jnm2** [comment]
@Lette Probably https://github.com/nunit/nunit/blob/master/src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs. Where is the UNIVERSE test?

But first I want to make sure that this makes sense to @nunit/framework-team.

I think it's weird that `Is.EqualTo(str)` is Ordinal and `Is.EqualTo(str).IgnoreCase` is CurrentCultureIgnoreCase. Seems it should be Ordinal and OrdinalIgnoreCase, or CurrentCulture and CurrentCultureIgnoreCase.

`Ordinal` is honestly probably the better default, contrary to the tests I posted above. We could add APIs for `Is.EqualTo(str).CurrentCulture` and `Is.EqualTo(str).CurrentCultureIgnoreCase`.

I'll wait for others to comment.

**Comment by Lette** [comment]
The `UNIVERSE` test is here: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/tests/Constraints/DictionaryContainsValueConstraintTests.cs#L87

**Comment by ChrisMaddock** [comment]
> I think it's weird that Is.EqualTo(str) is Ordinal and Is.EqualTo(str).IgnoreCase is CurrentCultureIgnoreCase. Seems it should be Ordinal and OrdinalIgnoreCase, or CurrentCulture and CurrentCultureIgnoreCase.

Hmm. Given we allow the user to set a Culture, I feel like the comparison should respect that, and use a CultureSensitive comparison. To change that now however feels pretty fundamental, and a bit of a subtle breaking change.

I agree they should be consistent. In an ideal world, I think we'd make them both culture sensitive, but I imagine changing the behaviour of the `IgnoreCase` version would affect significantly less users.

So in conclusion - I have no conclusion. 😄

**Comment by Lette** [comment]
Added the tests as requested. And as you suspected, the `tr-TR` version is now failing.

**Comment by jnm2** [comment]
@ChrisMaddock I guess it makes sense to keep it even if it's a bit arbitrary. Ordinal comparisons are the 99.9% case, but once you're ignoring case, I can see why culture becomes important. XML doc comments should clearly reference `<see cref="StringComparison.Ordinal"/>` and `<see cref="StringComparison.CurrentCultureIgnoreCase"/>` then.

I'm in favor of just these changes then:
 * Platform should compare using `OrdinalIgnoreCase` (or `InvariantCultureIgnoreCase`? but why?)
    - New test that fails before we make this change no matter what your system settings are.
      If @Lette would rather not, one of us can follow up in a later PR?
 * The UNIVERSE test should have `[SetCulture("iv")]` for invariant. (Or possibly `[SetCulture("en")]` because 'universe' is an English word, but I don't think that was the point of the test.)
 * The added tests, plus this to lock in the current ordinal behavior:
   ```
   [Test, SetCulture("iv")]
   public void Ordinal_comparer_used_by_default()
   {
       Assert.That("ss", Is.Not.EqualTo("ß"));
   }
   ```

File: src/NUnitFramework/framework/Internal/StringUtil.cs (hunk 2 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/StringUtil.cs]: @Lette My bad! `string.Compare` then...

File: src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs (hunk 3 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs]: @Lette Looks like we want to keep the original behavior, so let's not make any changes at all. `string.Equals(,,)` would be an improvement but orthogonal to the purposes of this PR.

File: src/NUnitFramework/framework/Internal/StringUtil.cs (hunk 2 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/StringUtil.cs]: Same [@Lette Looks like we want to keep the original behavior, so let's not make any changes at all. `string.Compare(,,)` would be an improvement but orthogonal to the purposes of this PR.]

**Comment by Lette** [comment]
The [docs](https://msdn.microsoft.com/en-us/library/system.globalization.cultureinfo.invariantculture(v=vs.110).aspx) state that:

> You specify the invariant culture by name by using an empty string ("") [...]

Where is `"iv"` documented?

**Comment by Lette** [comment]
Btw, I'm very happy to help. I'll clean up the PR according to the conclusions in this discussion.

**Comment by jnm2** [comment]
@Lette https://msdn.microsoft.com/en-us/library/system.globalization.cultureinfo.aspx#Invariant:

> An invariant culture is culture-insensitive. Your application specifies the invariant culture by name using an empty string ("") or by its identifier. 

I took `TwoLetterISOLanguageName` to be its identifier, but I'm not sure. `""` should be fine.

**Comment by jnm2** [comment]
@Lette 
> Btw, I'm very happy to help. I'll clean up the PR according to the conclusions in this discussion.

Great, thanks! Feel free to rebase if you like.

It'll be a little complex to add a test on the bad platform behavior that fails no matter what your system settings are. I think you'd have to add a dummy fixture in a new `PlatformFixture.cs` in `https://github.com/nunit/nunit/tree/master/src/NUnitFramework/testdata` which contains the original test that was failing, and then a new test in a new `PlatformAttributeTests.cs` in `https://github.com/nunit/nunit/tree/master/src/NUnitFramework/tests/Attributes` which uses `[SetCulture("tr")]` and checks that the `testdata` test runs properly using code similar to [this](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/tests/Attributes/TestFixtureSourceTests.cs#L50). Let us know if you get stuck.

**Comment by CharliePoole** [comment]
Very general comment and question - haven't done a real review...

1. It seems as if this is fixing a bunch of things that would probably be separate PRs if not for the fact that they all came up using the Turkish culture. Should we perhaps work on one area at a time?

2. Can the errors  be duplicated using SetCulture on a test or does it actually require changing the system culture? If SetCulture will work then we can write unit tests that fail.

**Comment by jnm2** [comment]
@CharliePoole 

1. Perhaps, but this is where where are; is it worth making @Lette split his PR into two?

2. My theory is that all we need is a SetCulture on a test which does a `TestBuilder.RunTest` on a test in `testdata` that has a `Platform` attribute. The tricky part will be that we will need to test both platform include `UNIX` and platform exclude `UNIX`, to observe the behavior whether or not you're on UNIX.
@Lette seriously, if you want me to take on this particular test, it's no problem.

**Comment by CharliePoole** [comment]
@jnm2 

1. The reason to do extra work is if we are not confident that we can adequately review what amounts to multiple logical changes in one PR. If we think we may break something, then going slower (i.e. smaller pieces) is worth it. The extra work for someone (not necessarily @Lette) pays for the assurance that we are not breaking anything for users. However, my remark was only based on a cursory examination. I'm happy to rely on the judgement of those doing a closer review.

2. Exactly.

**Comment by Lette** [comment]
@jnm2 , @ChrisMaddock , @CharliePoole 

Maybe you should count me out. I'll be travelling for the next week and I'll have limited coding time. If you can wait until next weekend with this, I can do it. Otherwise, you might be better off assigning someone else.

And, btw, fwiw,, I believe @jnm2 's theory regarding @CharliePoole 's second question is correct.

**Comment by ChrisMaddock** [comment]
@Lette - I don't think there's any rush, so long as you're still willing to look at it. 🙂

**Comment by Lette** [comment]
So, I'm restarting on this one.

Thought I'd start with the easy parts and promptly got stuck.

**Question:** Is it safe to say that `Wın` and `Unıx` (with Turkish small dotless i) and `WİN` and `UNİX (with Turkish large dotted I) are invalid platform names?

Right now, tests executed on Turkish Windows with `[Platform("Win")]` wont be run since `PlatformHelper.IsPlatformSupported` will do `.ToUpper()` on the name and get `WİN` which is not a match to a known platform. However, and for the same reason, tests with `[Platform("Wın")]` *will* be run.

If we "fix" this we might end up with having tests with `[Platform("Wın")]` *no longer run*... Could that be a problem?

**Comment by ChrisMaddock** [comment]
> If we "fix" this we might end up with having tests with [Platform("Wın")] no longer run... Could that be a problem?

I think that's a niche enough breaking change we can go with it! 🙂 My opinion would be to go for `Win` - the invariant string.

**Comment by jnm2** [comment]
I agree. I think InvariantIgnoreCase is exactly what we want. Platform names are trademarks and feel like things that should be spelled invariantly, in the same way that `while` and `async` are spelled invariantly in C#. I'd have a hard time believing this would break something.

IgnoreCase is important to me though because I'm almost sure we will break things if we require the correct capitalization. So more like VB.NET than C# in this regard. =)

**Comment by CharliePoole** [comment]
So let's fix PlatformHelper. The only reason it has ToUpper is that it was written to work on .NET 1.1!

**Comment by CharliePoole** [comment]
"If we only had the time" we could go through every class once every two or three years and remove the compromises due to previously supported platforms.

**Comment by ChrisMaddock** [comment]
Hey @Lette - are you still interested in coming back to this? 

I think it would be good to put in some fixes to these issues, but someone would need to go through each case, and make sure we're making the correct fix.

**Comment by ChrisMaddock** [comment]
Closing as this has gone stale. Happy to look at this again if anyone else wanted to pick it up.
